packages/gatsby-core-utils/src/mutex.ts

```diff
@@ -0,0 +1,53 @@
+import { getStorage, LockStatus, getDatabaseDir } from "./utils/get-storage"
+
+interface IMutex {
+  acquire(): Promise<void>
+  release(): Promise<void>
+}
+
+// Random number to re-check if mutex got released
+const MUTEX_INTERVAL = 3000
```

Comment:
This is probably arbitrarily chosen, any indication wether we want to make this a smaller or bigger number? I'd suppose at _scale_ some optimizations here would be possible

Reply:
For reference - this is for similar usage scenario (using lmdb to synchronise processes) (specific to secondary indexes) - https://github.com/gatsbyjs/gatsby/blob/ec20301df296aeff5ec29f23d21ddfadd40032c1/packages/gatsby/src/datastore/lmdb/query/create-index.ts#L224-L248

which has "expenential backoff". 

For file downloads it does make sense for it to be a bit higher, but here we have generic implementation (which will be used for file downloads in follow up).

I would consider making it possible to specify timeout for given mutex usage (with some default - might even be current `3000`) so depending on context and expectation on how long some "locked" work might take

Reply:
Added a timeout, I think backoff is going to be too slow